src/vision/status_quo/barbara_needs_async_helpers.md

```diff
@@ -8,28 +8,91 @@ If you would like to expand on this story, or adjust the answers to the FAQ, fee
+[Barbara], an experienced Rust user, is prototyping an async Rust service for work. To get things working quickly, she decides to prototype in [`tokio`], since it is unclear which runtime her work will use.
+
+She starts adding warp and [`tokio`] to her dependencies list. She notices that [`warp`] [suggests](https://github.com/seanmonstar/warp/#example) using [`tokio`] with the `full` feature. She's a bit concerned about how this might affect the compile times and also that *all* of tokio is needed for her little project, but she pushes forward.
+
+As she builds out functionality, she's pleased to see tokio provides a bunch of helpers like [`join!`](https://docs.rs/tokio/1.4.0/tokio/macro.join.html) and async versions of the standard library types like channels and mutexes.
+
+After completing one endpoint, she moves to a new one which requires streaming http responses to the client. Barbara quickly finds out from [`tokio`] [docs](https://docs.rs/tokio/1.4.0/tokio/stream/index.html), that it does not provide a stream type, and so she adds [`tokio-stream`] to her dependencies.
+
+Moving on she tries to make some functions generic over the web framework underneath, so she tries to abstract off the functionality to a trait. So she writes an async function inside a trait, just like a normal function.
+
+```rust
+trait Client {
+    async fn get();
+}
+```
+
+Then she gets a helpful error message.
+
+```
+error[E0706]: functions in traits cannot be declared `async`
+ --> src/lib.rs:2:5
+  |
+2 |     async fn get();
+  |     -----^^^^^^^^^^
+  |     |
+  |     `async` because of this
+  |
+  = note: `async` trait functions are not currently supported
+  = note: consider using the `async-trait` crate: https://crates.io/crates/async-trait
+```
+
+She then realizes that Rust doesn't support async functions in traits yet, so she adds [`async-trait`] to her dependencies.
+
+Some of her functions are recursive, and she wanted them to be async functions, so she sprinkles some `async/.await` keywords in those functions.
+
+```rust
+async fn sum(n: usize) -> usize {
+    if n == 0 {
+        0
+    } else {
+        n + sum(n - 1).await
+    }
+}
+```
+
+Then she gets an error message.
+
+```
+error[E0733]: recursion in an `async fn` requires boxing
+ --> src/lib.rs:1:27
+  |
+1 | async fn sum(n: usize) -> usize {
+  |                           ^^^^^ recursive `async fn`
+  |
+  = note: a recursive `async fn` must be rewritten to return a boxed `dyn Future`
+```
+
+So to make these functions async she starts boxing her futures. She also asks one of her peers for a code review asynchronously, and after awaiting their response, she learns about the [`async-recursion`] crate. Then she adds [`async-recursion`] to the dependencies.
```

Comment:
> So to make these functions async she starts boxing her futures.

It'd be nice to see a code example of what this looks like

Reply:
@rylev I think you said you were boxing your futures like a barbarian, maybe you can provide an example?

Reply:
> > So to make these functions async she starts boxing her futures.
> 
> It'd be nice to see a code example of what this looks like

What about something like this?

---

She tries boxing her futures, the hard way, fighting with the compiler. She knows that `async` keyword is sort of a sugar for `impl Future` so she tries the following at first.

```rust
fn sum(n: usize) -> Box<dyn Future<Output = usize>> {
    Box::new(async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    })
}
```

The compiler gives an error.

```
error[E0277]: `dyn Future<Output = usize>` cannot be unpinned
  --> src/main.rs:11:17
   |
11 |             n + sum(n - 1).await
   |                 ^^^^^^^^^^^^^^^^ the trait `Unpin` is not implemented for `dyn Future<Output = usize>`
   |
   = note: required because of the requirements on the impl of `Future` for `Box<dyn Future<Output = usize>>`
   = note: required by `poll`
```

She then reads about `Unpin` and `Pin`, and then finally comes up with a solution.

```rust
fn sum(n: usize) -> Pin<Box<dyn Future<Output = usize>>> {
    Box::pin(async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    })
}
```

The code works!

---

I assumed here that she would take a route without involving the `futures` crate. With the `futures` crate, it is as easy as:

```rust
use futures::future::{BoxFuture, FutureExt};

fn sum(n: usize) -> BoxFuture<'static, usize> {
    async move {
        if n == 0 {
            0
        } else {
            n + sum(n - 1).await
        }
    }.boxed()
}
```

Reply:
that looks great, thanks. As I expected, it's a significant change -- she has to move from `async fn` to a `fn`. In this case, the signature is fairly easy, but it can be quite a bit more difficult if regions are involved.

Reply:
I would probably have her write it by hand then later learn about futures crate, then about async-recursion.

Reply:
This looks great!

Reply:
> that looks great, thanks. As I expected, it's a significant change -- she has to move from `async fn` to a `fn`. In this case, the signature is fairly easy, but it can be quite a bit more difficult if lifetimes are involved.

Is an example with lifetimes also required?

Reply:
Nah, I don't think that's necessary. You could add a FAQ about it though.